artemis-server/src/main/java/org/apache/activemq/artemis/core/server/impl/ServerSessionImpl.java

```diff
@@ -1742,49 +1743,80 @@ public Transaction getCurrentTransaction() {
+            try {
+               createAddress(queueConfig.getAddress(), queueConfig.getRoutingType(), true).setTemporary(queueConfig.isTemporary());
+            } catch (ActiveMQAddressExistsException e) {
+               // The address may have been created by another thread in the mean time.  Catch and do nothing.
+            }
+            result = AutoCreateResult.CREATED;
+         } else {
+            // if the address doesn't exist and can't be autocreated then return false immediately
+            return AutoCreateResult.NOT_FOUND;
+         }
+      } else {
+         // the address exists
+         if (addressInfo.getRoutingTypes().contains(queueConfig.getRoutingType())) {
+            // the existing address supports the requested routingType
+            result = AutoCreateResult.EXISTED;
+         } else {
+            // the existing address doesn't support the requested routingType
+            if (addressSettings.isAutoCreateAddresses() || queueConfig.isTemporary()) {
+               // try to update the address with the new routingType if possible
                try {
-                  createAddress(address, routingType, true);
+                  createAddress(addressInfo.addRoutingType(queueConfig.getRoutingType()), true);
                } catch (ActiveMQAddressExistsException e) {
                   // The address may have been created by another thread in the mean time.  Catch and do nothing.
                }
-               result = true;
+               result = AutoCreateResult.UPDATED;
             } else {
-               result = false;
+               // if the address exists but doesn't support the requested routingType and can't be updated with the new routingType then return false immediately
+               return AutoCreateResult.NOT_FOUND;
             }
-         } else {
-            result = true;
          }
-      } else if (routingType == RoutingType.ANYCAST) {
+      }
+
+      if (queueConfig.getRoutingType() == RoutingType.ANYCAST) {
          if (server.locateQueue(unPrefixedAddress) == null) {
-            Bindings bindings = server.getPostOffice().lookupBindingsForAddress(address);
+            // the queue doesn't exist
+            Bindings bindings = server.getPostOffice().lookupBindingsForAddress(queueConfig.getAddress());
             if (bindings != null) {
                // this means the address has another queue with a different name, which is fine, we just ignore it on this case
-               result = true;
-            } else if (addressSettings.isAutoCreateQueues()) {
+               result = AutoCreateResult.EXISTED;
+            } else if (addressSettings.isAutoCreateQueues() || queueConfig.isTemporary()) {
+               // try to create the queue
                try {
-                  createQueue(new QueueConfiguration(address).setRoutingType(routingType).setAutoCreated(true));
+                  createQueue(queueConfig.setAutoCreated(true));
```

Comment:
manipulating the passed in QueueConfig feels a little icky. If doing so, should the method doc that callers of it shouldnt reuse them?